File: animax/platform/android/animax_android/src/main/java/com/lynx/animax/property/AnimaXValueParam.java
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynx.animax.property;

import com.lynx.animax.base.CalledByNative;
import java.lang.Number;

/**
 * A value parameter class that can hold different types of values (String, Number, Boolean,
 * Coordinate) for animation parameters. Supports optional target frame for keyframe animations.
 */
public class AnimaXValueParam {
  public static final int FRAME_INDEX_ALL = -1;

  /**
   * A simple class to hold 3D coordinate data (x, y, z).
   */
  private static class Coordinate {
    public final double x;
    public final double y;
    public final double z;

    public Coordinate(double x, double y, double z) {
      this.x = x;
      this.y = y;
      this.z = z;
    }
  }

  private enum Type { STRING, NUMBER, BOOLEAN, COORDINATE, COLOR, COLOR_FILTER, NULL }

  private final Type mType;
  private final String mStringValue;
  private final Number mNumberValue;
  private final Boolean mBooleanValue;
  private final Coordinate mCoordinateValue;
  private final Integer mColorValue;
  private final Integer mTargetFrame;

  private AnimaXValueParam(Type type, String stringValue, Number numberValue, Boolean booleanValue,
      Coordinate coordinateValue, Integer colorValue, Integer targetFrame) {
    this.mType = type;
    this.mStringValue = stringValue;
    this.mNumberValue = numberValue;
    this.mBooleanValue = booleanValue;
    this.mCoordinateValue = coordinateValue;
    this.mColorValue = colorValue;
    this.mTargetFrame = targetFrame;
  }

  /**
   * Creates a AnimaXValueParam from a String value
   *
   * @param value String value
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromString(String value) {
    return new AnimaXValueParam(Type.STRING, value, null, null, null, null, null);
  }

  /**
   * Creates a AnimaXValueParam from a String value with a target frame
   *
   * @param value String value
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromString(String value, int targetFrame) {
    return new AnimaXValueParam(Type.STRING, value, null, null, null, null, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam from a numeric value (int, float, double)
   *
   * @param value Numeric value
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromNumber(double value) {
    return new AnimaXValueParam(Type.NUMBER, null, value, null, null, null, null);
  }

  /**
   * Creates a AnimaXValueParam from a numeric value with a target frame
   *
   * @param value Numeric value
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromNumber(double value, int targetFrame) {
    return new AnimaXValueParam(Type.NUMBER, null, value, null, null, null, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam from a boolean value
   *
   * @param value Boolean value
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromBoolean(boolean value) {
    return new AnimaXValueParam(Type.BOOLEAN, null, null, value, null, null, null);
  }

  /**
   * Creates a AnimaXValueParam from a boolean value with a target frame
   *
   * @param value Boolean value
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromBoolean(boolean value, int targetFrame) {
    return new AnimaXValueParam(Type.BOOLEAN, null, null, value, null, null, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam from a 2D coordinate (x,y coordinates)
   *
   * @param x X coordinate
   * @param y Y coordinate
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromCoordinate(double x, double y) {
    return new AnimaXValueParam(
        Type.COORDINATE, null, null, null, new Coordinate(x, y, 0), null, null);
  }

  /**
   * Creates a AnimaXValueParam from a 2D coordinate with a target frame
   *
   * @param x X coordinate
   * @param y Y coordinate
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromCoordinate(double x, double y, int targetFrame) {
    return new AnimaXValueParam(
        Type.COORDINATE, null, null, null, new Coordinate(x, y, 0), null, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam from a 3D coordinate (x,y,z coordinates)
   *
   * @param x X coordinate
   * @param y Y coordinate
   * @param z Z coordinate
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromCoordinate(double x, double y, double z) {
    return new AnimaXValueParam(
        Type.COORDINATE, null, null, null, new Coordinate(x, y, z), null, null);
  }

  /**
   * Creates a AnimaXValueParam from a 3D coordinate with a target frame
   *
   * @param x X coordinate
   * @param y Y coordinate
   * @param z Z coordinate
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromCoordinate(double x, double y, double z, int targetFrame) {
    return new AnimaXValueParam(
        Type.COORDINATE, null, null, null, new Coordinate(x, y, z), null, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam from a color value
   *
   * @param color Color integer value
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromColor(int color) {
    return new AnimaXValueParam(Type.COLOR, null, null, null, null, color, null);
  }

  /**
   * Creates a AnimaXValueParam from a color value with a target frame
   *
   * @param color Color integer value
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromColor(int color, int targetFrame) {
    return new AnimaXValueParam(Type.COLOR, null, null, null, null, color, targetFrame);
  }

  /**
   * Creates a AnimaXValueParam for a color filter with color and mode
   *
   * @param color Color integer value
   * @param mode Filter mode value
   * @return AnimaXValueParam instance
   */
  @CalledByNative
  public static AnimaXValueParam fromColorFilter(int color, int mode) {
    return new AnimaXValueParam(Type.COLOR_FILTER, null, (double) mode, null, null, color, null);
  }

  /**
   * Creates a AnimaXValueParam for a color filter with color and mode, and a target frame
   *
   * @param color Color integer value
   * @param mode Filter mode value
   * @param targetFrame Frame number at which this value should be applied
   * @return AnimaXValueParam instance
   */
  public static AnimaXValueParam fromColorFilter(int color, int mode, int targetFrame) {
    return new AnimaXValueParam(
        Type.COLOR_FILTER, null, (double) mode, null, null, color, targetFrame);
  }

  /**
   * Creates a null AnimaXValueParam
   *
   * @return AnimaXValueParam instance with null value
   */
  public static AnimaXValueParam createNull() {
    return new AnimaXValueParam(Type.NULL, null, null, null, null, null, null);
  }

  /**
   * Checks if this parameter has a target frame
   *
   * @return true if target frame is specified
   */
  public boolean hasTargetFrame() {
    return mTargetFrame != null && mTargetFrame != FRAME_INDEX_ALL;
  }

  /**
   * Gets the target frame if specified
   *
   * @return Target frame or -1 if not specified
   */
  @CalledByNative
  public int getTargetFrame() {
    return mTargetFrame != null ? mTargetFrame : FRAME_INDEX_ALL;
  }

  /**
   * Checks if the value is a String
   *
   * @return true if value is a String
   */
  public boolean isString() {
    return mType == Type.STRING;
  }

  /**
   * Checks if the value is a Number
   *
   * @return true if value is a Number
   */
  public boolean isNumber() {
    return mType == Type.NUMBER;
  }

  /**
   * Checks if the value is a Boolean
   *
   * @return true if value is a Boolean
   */
  public boolean isBoolean() {
    return mType == Type.BOOLEAN;
  }

  /**
   * Checks if the value is a Coordinate
   *
   * @return true if value is a Coordinate
   */
  public boolean isCoordinate() {
    return mType == Type.COORDINATE;
  }

  /**
   * Checks if the value is a Color
   *
   * @return true if value is a Color
   */
  public boolean isColor() {
    return mType == Type.COLOR;
  }

  /**
   * Checks if the value is a ColorFilter
   *
   * @return true if value is a ColorFilter
   */
  public boolean isColorFilter() {
    return mType == Type.COLOR_FILTER;
  }

  /**
   * Checks if the value is null
   *
   * @return true if value is null
   */
  public boolean isNull() {
    return mType == Type.NULL;
  }

  /**
   * Gets the value as a String
   *
   * @return String value or null if not a String
   */
  @CalledByNative
  public String getStringValue() {
    return mStringValue;
  }

  /**
   * Gets the value as a Number
   *
   * @return Number value or null if not a Number
   */
  @CalledByNative
  public double getNumberValue() {
    return mNumberValue != null ? mNumberValue.doubleValue() : 0.0;
  }

  /**
   * Gets the value as a Boolean
   *
   * @return Boolean value or null if not a Boolean
   */
  @CalledByNative
  public boolean getBooleanValue() {
    return mBooleanValue != null ? mBooleanValue : false;
  }

  /**
   * Gets the X coordinate if this is a Coordinate value
   *
   * @return X coordinate or 0 if not a Coordinate
   */
  @CalledByNative
  public double getX() {
    return mCoordinateValue != null ? mCoordinateValue.x : 0d;
  }

  /**
   * Gets the Y coordinate if this is a Coordinate value
   *
   * @return Y coordinate or 0 if not a Coordinate
   */
  @CalledByNative
  public double getY() {
    return mCoordinateValue != null ? mCoordinateValue.y : 0d;
  }

  /**
   * Gets the Z coordinate if this is a Coordinate value
   *
   * @return Z coordinate or 0 if not a Coordinate
   */
  @CalledByNative
  public double getZ() {
    return mCoordinateValue != null ? mCoordinateValue.z : 0d;
  }

  /**
   * Gets the color value
   *
   * @return Color int value or 0 if not a Color or ColorFilter
   */
  @CalledByNative
  public int getColorValue() {
    if (isColor() || isColorFilter()) {
      return mColorValue != null ? mColorValue : 0;
    }
    return 0;
  }

  /**
   * Gets the filter mode value if this is a ColorFilter
   *
   * @return Mode value or 0 if not a ColorFilter
   */
  @CalledByNative
  public int getFilterMode() {
    return isColorFilter() ? (int) getNumberValue() : 0;
  }

  @CalledByNative
  public int getTypeIndex() {
    return mType.ordinal();
  }

  @Override
  public String toString() {
    int targetFrame = getTargetFrame();
    switch (mType) {
      case STRING:
        return "AnimaXValueParam{stringValue='" + mStringValue + "', targetFrame=" + targetFrame
            + "}";
      case NUMBER:
        return "AnimaXValueParam{numberValue=" + getNumberValue() + ", targetFrame=" + targetFrame
            + "}";
      case BOOLEAN:
        return "AnimaXValueParam{booleanValue=" + getBooleanValue() + ", targetFrame=" + targetFrame
            + "}";
      case COORDINATE:
        return "AnimaXValueParam{coordinateValue=(" + getX() + "," + getY() + "," + getZ()
            + "), targetFrame=" + targetFrame + "}";
      case COLOR:
        return "AnimaXValueParam{colorValue=" + getColorValue() + ", targetFrame=" + targetFrame
            + "}";
      case COLOR_FILTER:
        return "AnimaXValueParam{colorValue=" + getColorValue() + ", mode=" + getNumberValue()
            + ", targetFrame=" + targetFrame + "}";
      case NULL:
        return "AnimaXValueParam{null, targetFrame=" + targetFrame + "}";
      default:
        return "AnimaXValueParam{unknown}";
    }
  }
}

File: src/property/animax_value_param.cc
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "include/property/animax_value_param.h"

#include "src/model/value/base_value.h"
#include "src/model/value/document_data.h"
#include "src/render/color_filter.h"

namespace lynx {
namespace animax {

namespace {
ValueType GetValueTypeForProperty(LayerPropertyType property_type) {
  switch (property_type) {
    case LayerPropertyType::kTransformRotation:
    case LayerPropertyType::kTransformSkew:
    case LayerPropertyType::kTransformSkewAngle:
    case LayerPropertyType::kTransformStartOpacity:
    case LayerPropertyType::kTransformEndOpacity:
    case LayerPropertyType::kTransformRotationX:
    case LayerPropertyType::kTransformRotationY:
    case LayerPropertyType::kTransformRotationZ:
    case LayerPropertyType::kTextSize:
    case LayerPropertyType::kTextTracking:
    case LayerPropertyType::kTextSkew:
    case LayerPropertyType::kTextRangeOffset:
    case LayerPropertyType::kTextRangeStart:
    case LayerPropertyType::kTextRangeEnd:
    case LayerPropertyType::kStrokeWidth:
    case LayerPropertyType::kBlurRadius:
    case LayerPropertyType::kCornerRadius:
    case LayerPropertyType::kRepeaterCopies:
    case LayerPropertyType::kPolyStarPoints:
    case LayerPropertyType::kPolyStarRotation:
    case LayerPropertyType::kPolyStarInnerRadius:
    case LayerPropertyType::kPolyStarOuterRadius:
    case LayerPropertyType::kPolyStarInnerRounded:
    case LayerPropertyType::kPolyStarOuterRounded:
    case LayerPropertyType::kDropShadowDirection:
    case LayerPropertyType::kDropShadowDistance:
    case LayerPropertyType::kDropShadowRadius:
      return ValueType::kFloat;

    case LayerPropertyType::kVisibility:
    case LayerPropertyType::kTransformOpacity:
    case LayerPropertyType::kOpacity:
    case LayerPropertyType::kDropShadowOpacity:
      return ValueType::kInteger;

    case LayerPropertyType::kTransformPosition:
    case LayerPropertyType::kTransformAnchor:
    case LayerPropertyType::kPosition:
    case LayerPropertyType::kEllipseSize:
    case LayerPropertyType::kRectangleSize:
    case LayerPropertyType::kRepeaterOffset:
      return ValueType::kPoint;

    case LayerPropertyType::kTransformScale:
      return ValueType::kScale;

    case LayerPropertyType::kColor:
    case LayerPropertyType::kTextColor:
    case LayerPropertyType::kStrokeColor:
    case LayerPropertyType::kDropShadowColor:
      return ValueType::kColor;

    case LayerPropertyType::kColorFilter:
      return ValueType::kColorFilter;

    case LayerPropertyType::kTextValue:
      return ValueType::kDocument;

    default:
      return ValueType::kUnknown;
  }
}
}  // namespace

AnimaXValueParam::AnimaXValueParam(const std::string& value,
                                   int32_t target_frame)
    : type_(Type::kString), string_value_(value), target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(double value, ApplyMode apply_mode,
                                   int32_t target_frame)
    : type_(Type::kNumber),
      number_value_(value),
      apply_mode_(apply_mode),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(bool value, int32_t target_frame)
    : type_(Type::kBoolean),
      number_value_(value ? 1.0 : 0.0),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(double x, double y, ApplyMode apply_mode,
                                   int32_t target_frame)
    : type_(Type::kCoordinate),
      coordinate_value_(PointF(static_cast<float>(x), static_cast<float>(y))),
      apply_mode_(apply_mode),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(double x, double y, double z,
                                   ApplyMode apply_mode, int32_t target_frame)
    : type_(Type::kCoordinate),
      coordinate_value_(PointF(static_cast<float>(x), static_cast<float>(y),
                               static_cast<float>(z))),
      apply_mode_(apply_mode),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(int32_t color, int32_t target_frame)
    : type_(Type::kColor),
      color_value_(Color(color)),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam(const ColorFilterParam& filter_params,
                                   int32_t target_frame)
    : type_(Type::kColorFilter),
      number_value_(static_cast<double>(filter_params.mode)),
      color_value_(Color(filter_params.color)),
      target_frame_(target_frame) {}

AnimaXValueParam::AnimaXValueParam()
    : type_(Type::kNull), target_frame_(kFrameIndexAll) {}

AnimaXValueParam& AnimaXValueParam::operator=(
    AnimaXValueParam&& other) noexcept {
  if (this != &other) {
    type_ = other.type_;
    string_value_ = std::move(other.string_value_);
    number_value_ = std::move(other.number_value_);
    coordinate_value_ = std::move(other.coordinate_value_);
    color_value_ = std::move(other.color_value_);
    apply_mode_ = other.apply_mode_;
    target_frame_ = other.target_frame_;
  }
  return *this;
}

const std::string& AnimaXValueParam::GetStringValue() const {
  if (string_value_.has_value()) {
    return *string_value_;
  }
  static const std::string kEmptyString = "";
  return kEmptyString;
}

double AnimaXValueParam::GetNumberValue() const {
  return number_value_.value_or(0.0);
}

bool AnimaXValueParam::GetBooleanValue() const {
  return number_value_.value_or(0.0) == 1.0;
}

double AnimaXValueParam::GetX() const {
  return coordinate_value_ ? coordinate_value_->GetX() : 0.0;
}

double AnimaXValueParam::GetY() const {
  return coordinate_value_ ? coordinate_value_->GetY() : 0.0;
}

double AnimaXValueParam::GetZ() const {
  return coordinate_value_ ? coordinate_value_->GetZ() : 0.0;
}

int32_t AnimaXValueParam::GetColorValue() const {
  if ((IsColor() || IsColorFilter()) && color_value_) {
    return color_value_->GetInt();
  }
  return 0;
}

void* AnimaXValueParam::GetValuePtr() const {
  switch (type_) {
    case Type::kString:
      return string_value_.has_value()
                 ? const_cast<std::string*>(&*string_value_)
                 : nullptr;
    case Type::kNumber:
    case Type::kBoolean:
      return number_value_.has_value() ? const_cast<double*>(&*number_value_)
                                       : nullptr;
    case Type::kCoordinate:
      return coordinate_value_.has_value()
                 ? const_cast<PointF*>(&*coordinate_value_)
                 : nullptr;
    case Type::kColor:
      return color_value_.has_value() ? const_cast<Color*>(&*color_value_)
                                      : nullptr;
    // Do not support kColorFilter
    case Type::kNull:
    default:
      return nullptr;
  }
}

std::unique_ptr<AnimaXValueParam> AnimaXValueParam::FromValue(
    const Value* value) {
  if (!value) {
    return std::make_unique<AnimaXValueParam>();
  }

  auto value_type = value->GetValueType();
  switch (value_type) {
    case ValueType::kFloat: {
      auto* float_value = static_cast<const Float*>(value);
      return std::make_unique<AnimaXValueParam>(float_value->Get(),
                                                ApplyMode::kSet);
    }

    case ValueType::kInteger: {
      auto* int_value = static_cast<const Integer*>(value);
      return std::make_unique<AnimaXValueParam>(
          static_cast<double>(int_value->Get()), ApplyMode::kSet);
    }

    case ValueType::kPoint: {
      auto* point_value = static_cast<const PointF*>(value);
      return std::make_unique<AnimaXValueParam>(
          point_value->GetX(), point_value->GetY(), point_value->GetZ(),
          ApplyMode::kSet);
    }

    case ValueType::kScale: {
      auto* scale_value = static_cast<const ScaleF*>(value);
      return std::make_unique<AnimaXValueParam>(
          scale_value->GetX(), scale_value->GetY(), scale_value->GetZ(),
          ApplyMode::kSet);
    }

    case ValueType::kColor: {
      auto* color_value = static_cast<const Color*>(value);
      return std::make_unique<AnimaXValueParam>(color_value->GetInt());
    }

    case ValueType::kColorFilter: {
      auto* color_filter_value = static_cast<const ColorFilter*>(value);
      return std::make_unique<AnimaXValueParam>(color_filter_value->GetColor(),
                                                color_filter_value->GetMode());
    }

    case ValueType::kDocument: {
      auto* document_value = static_cast<const DocumentData*>(value);
      return std::make_unique<AnimaXValueParam>(document_value->GetText());
    }

    case ValueType::kUnknown:
    default:
      return std::make_unique<AnimaXValueParam>();
  }
}

/**
 * Converts AnimaXValueParam to appropriate Value based on property type
 */
std::unique_ptr<Value> AnimaXValueParam::ToValue(
    LayerPropertyType property_type) const {
  ValueType value_type = GetValueTypeForProperty(property_type);
  if (value_type == ValueType::kUnknown) {
    return nullptr;
  }

  // Convert based on target value type
  switch (value_type) {
    case ValueType::kFloat:
      if (IsNumber()) {
        return ValueFactory::Make<Float>(static_cast<float>(GetNumberValue()));
      }
      break;

    case ValueType::kInteger:
      if (IsNumber()) {
        return ValueFactory::Make<Integer>(
            static_cast<int32_t>(GetNumberValue()));
      }
      break;

    case ValueType::kPoint:
      if (IsCoordinate()) {
        return ValueFactory::Make<PointF>(static_cast<float>(GetX()),
                                          static_cast<float>(GetY()),
                                          static_cast<float>(GetZ()));
      }
      break;

    case ValueType::kScale:
      if (IsCoordinate()) {
        return ValueFactory::Make<ScaleF>(static_cast<float>(GetX()),
                                          static_cast<float>(GetY()),
                                          static_cast<float>(GetZ()));
      }
      break;

    case ValueType::kColor:
      if (IsColor()) {
        return ValueFactory::Make<Color>(GetColorValue());
      }
      break;

    case ValueType::kColorFilter:
      if (IsColorFilter()) {
        return ValueFactory::Make<ColorFilter>(GetColorValue(),
                                               GetNumberValue());
      }
      break;

    case ValueType::kDocument:
      if (IsString()) {
        auto document_data = ValueFactory::Make<DocumentData>();
        document_data->SetText(GetStringValue());
        return document_data;
      }
      break;

    default:
      break;
  }

  return nullptr;
}

AnimaXValueParam::Type AnimaXValueParam::GetType() const { return type_; }

bool AnimaXValueParam::CopyTo(std::string* target) const {
  if (!target || !IsString() || !string_value_.has_value()) {
    return false;
  }
  *target = *string_value_;
  return true;
}

bool AnimaXValueParam::CopyTo(float* target) const {
  if (!target || !IsNumber() || !number_value_.has_value()) {
    return false;
  }
  *target = static_cast<float>(*number_value_);
  return true;
}

bool AnimaXValueParam::CopyTo(int32_t* target) const {
  if (!target || !IsNumber() || !number_value_.has_value()) {
    return false;
  }
  *target = static_cast<int32_t>(*number_value_);
  return true;
}

bool AnimaXValueParam::CopyTo(double* target) const {
  if (!target || !IsNumber() || !number_value_.has_value()) {
    return false;
  }
  *target = *number_value_;
  return true;
}

bool AnimaXValueParam::CopyTo(bool* target) const {
  if (!target || !IsBoolean() || !number_value_.has_value()) {
    return false;
  }
  *target = (*number_value_ == 1.0);
  return true;
}

}  // namespace animax
}  // namespace lynx
